event-web-framework/draft.markdown

```diff
@@ -0,0 +1,872 @@
+	        collect (cons (->keyword name) (or val ""))))
+```
+
+In the `parse` method specializing on `string`, we transform the content into usable pieces. We do this on strings instead of working directly with buffers because this makes it easier to test the actual parsing code in an environment like an interpreter or REPL. 
+
+The parsing process is:
+
+1. split on `"\\r?\\n"`
+2. split the first line of that on `" "` to get the request type (`POST`, `GET`, etc)/URI path/http-version
+3. assert that we're dealing with an `HTTP/1.1` request
+4. split the URI path on `"?"`, which gives us plain resource separate from any potential `GET` parameters
+5. make a new `request` instance with the resource in place
+6. populate that `request` instance with each split header line
+7. set that `request`s parameters to the result of parsing our `GET` parameters
+
+As you might expect by now, `request` is an instance of a CLOS class:
+
+```lisp
+	(defclass request ()
+	  ((resource :accessor resource :initarg :resource)
+	   (headers :accessor headers :initarg :headers :initform nil)
+	   (parameters :accessor parameters :initarg :parameters :initform nil)))
+```
+
+We've now seen how our clients can send requests and have them interpreted and handled by our server. The last thing we have to implement as part of our core server interface is the capability to write responses back to the client.
+
+### Rendering Responses
+
+Before we discuss rendering responses, we have to consider that there are two kinds of responses that we may be returning to our clients. The first is a "normal" HTTP response, complete with HTTP headers and body. We represent these kinds of responses with instances of the `response` class:
+
+```lisp
+	(defclass response ()
+	  ((content-type :accessor content-type :initform "text/html" :initarg :content-type)
+	   (charset :accessor charset :initform "utf-8")
+	   (response-code :accessor response-code :initform "200 OK" :initarg :response-code)
+	   (keep-alive? :accessor keep-alive? :initform nil :initarg :keep-alive?)
+	   (body :accessor body :initform nil :initarg :body)))
+```
+
+The second is an [SSE message](http://www.w3.org/TR/eventsource/), which we will use to send an incremental update to our clients.
+
+```lisp
+	(defclass sse ()
+	  ((id :reader id :initarg :id :initform nil)
+	   (event :reader event :initarg :event :initform nil)
+       (retry :reader retry :initarg :retry :initform nil)
+	   (data :reader data :initarg :data)))
+```
+
+We'll send an HTTP response whenever we receive a full HTTP request; however, how do we know when and where to send SSE messages without an originating client request?
+
+A simple solution is to register _channels_, to which we'll subscribe `socket`s as necessary.
+
+```lisp
+	(defparameter *channels* (make-hash-table))
+
+	(defmethod subscribe! ((channel symbol) (sock usocket))
+	  (push sock (gethash channel *channels*))
+	  nil)
+```
+
+NOTE: What is defparameter?
```

Comment:
The CL equivalent of `var`. It starts a variable definition form. The first symbol following it will be the name of the variable, and the second will be its value.